### PR TITLE
Remove dependency on shared heap from frametables.

### PR DESCRIPTION
**Pull request by stedolan**
Currently, frametables are allocated on the shared GC heap, to ensure that they get freed. This leads to a couple of "how did that ever work" circularities during domain initialisation.

With this patch, frametables are now statically allocated. Deallocation of old frametables is done by a check during stack scanning, and no frametables are deallocated until after the end of the GC cycle in which they were replaced. (This is effectively RCU).

### CODE REVIEW

File: asmrun/frame_descriptors.c [hunk 1]
```diff
@@ -81,48 +98,70 @@ static link *cons(intnat *frametable, link *tl) {
...
 
   caml_plat_unlock(&descr_mutex);
 }
 
-frame_descr* caml_find_frame_descr(uintnat pc)
+caml_frame_descrs caml_get_frame_descrs()
+{
+  struct frametable_version *ft =
+    (struct frametable_version*)atomic_load_acq(&current_frametable);
+  Assert(ft);
```
Comment (kayceesrk) [inline asmrun/frame_descriptors.c]: `CAMLassert` instead of `Assert`?

File: asmrun/frame_descriptors.c [hunk 2]
```diff
@@ -1,8 +1,12 @@
 #define CAML_INTERNALS
 
 #include "frame_descriptors.h"
-#include "caml/stack.h"
 #include "caml/platform.h"
+#include "caml/major_gc.h" /* for caml_major_cycles_completed */
+#include <stddef.h>
```
Comment (kayceesrk) [inline asmrun/frame_descriptors.c]: Why do you need this?

Comment (stedolan) [reply asmrun/frame_descriptors.c]: for `NULL`

File: asmrun/frame_descriptors.c (hunk 1 above)
Comment (stedolan) [reply asmrun/frame_descriptors.c]: Fixed

File: asmrun/frame_descriptors.c (hunk 2 above)
Comment (kayceesrk) [reply asmrun/frame_descriptors.c]: cool.

**Comment by stedolan** [comment]
The Travis failure is real, but it's a rare nondeterministic bug and seems to have nothing to do with this commit (I can reproduce it on master). Debugging atm, I'll make a new PR when I figure it out.

**Pull request merged**